packages/context/src/context.ts

```diff
@@ -211,18 +265,22 @@ export class Context {
    * binding is found, an error will be thrown.
    *
    * @param key Binding key
+   * @param options Options to control if the binding is optional. If
+   * `options.optional` is set to true, the method will return `undefined`
+   * instead of throwing an error if the binding key is not found.
    */
-  getBinding(key: string): Binding {
+  getBinding(key: string, options?: {optional?: boolean}): Binding | undefined {
```

Comment:
I am little bit concerned that the new signature will require many callers to explicitly handle the case of `getBinding()` returning `undefined` even when they don't provide any options.

Consider the following snippet:

```ts
const b: Binding = ctx.getBinding('foo');
```

Before this change, the compiler will happily accept this line of code.

After your change, I am expecting the compiler to complain that it cannot assign `Binding | undefined` to `Binding`.

I am proposing to improve the type definition for this method to explicitly tell the compiler that we always return `Binding` when there are no `options` argument. I hope something like the following is possible:

```ts
getBinding(key: string): Binding;
getBinding(key: string, options: {optional?: boolean}): Binding | undefined;

getBinding(key: string, options?: {optional?: boolean}): Binding | undefined {
  // the implementation
}
```

We may want to take this to the next level by including different `options.optional` values in the type signatures, but that may be overkill and/or not work as intended because of limitations of TypeScript.

```ts
interface GetBindingOptions {
  optional?: boolean;
}

getBinding(key: string): Binding;
getBinding(key: string, GetBindingOptions & {optional: false | undefined}): Binding;
getBinding(key: string, GetBindingOptions & {optional: true}): Binding | undefined;

getBinding(key: string, options?: GetBindingOptions): Binding | undefined { 
  // the implementation
}
```

Reply:
Ironically, `conditional types` is coming in TypeScript 2.8.

For now, the caller can use `!` to tell `tsc` it's not undefined. For example:
```ts
const binding = ctx.getBinding('x')!;
```

Reply:
I know users can use `!`, but that defeats the purpose of strict-null-check performed by the compiler, right?

My first proposal seems to work exactly as I thought, I added a commit to your feature branch to show what I mean - see 46794e0c

@raymondfeng please let me know what do you think about that change.

Reply:
👍